File: Anul 2/Semestrul 2/Metode de proiectare si programare/labs/Tema 3-proiect Server-Client Java/proiectServerClientJava/Server/src/main/java/proiectServerClientJava/server/Service.java
package proiectServerClientJava.server;

import proiectServerClientJava.domain.Angajat;
import proiectServerClientJava.domain.Bilet;
import proiectServerClientJava.domain.Spectacol;
import proiectServerClientJava.domain.validators.ValidationException;
import proiectServerClientJava.domain.validators.Validator;
import proiectServerClientJava.persistence.repository.jdbc.RepoAngajat;
import proiectServerClientJava.persistence.repository.jdbc.RepoBilet;
import proiectServerClientJava.persistence.repository.jdbc.RepoSpectacol;
import proiectServerClientJava.service.IObserver;
import proiectServerClientJava.service.IService;
import proiectServerClientJava.service.MyException;

import java.time.LocalTime;
import java.util.Map;
import java.util.Objects;
import java.util.concurrent.ConcurrentHashMap;

public class Service implements IService{
    private final Validator<Angajat> angajatValidator;
    private final Validator<Spectacol> spectacolValidator;
    private final Validator<Bilet> biletValidator;
    private final RepoAngajat angajatRepo;
    private final RepoSpectacol spectacolRepo;
    private final RepoBilet biletRepo;
    private Map<Long, IObserver> loggedAngajati;


    public Service(Validator<Angajat> validatorA,
                   Validator<Spectacol> validatorS,
                   Validator<Bilet> validatorB,
                   RepoAngajat angajatRepo,
                   RepoSpectacol spectacolRepo,
                   RepoBilet biletRepo)
    {
        this.angajatValidator=validatorA;
        this.spectacolValidator=validatorS;
        this.biletValidator=validatorB;
        this.angajatRepo=angajatRepo;
        this.spectacolRepo=spectacolRepo;
        this.biletRepo=biletRepo;
        loggedAngajati =new ConcurrentHashMap<>();
    }

    @Override
    public Angajat getAngajatByUsername(String username){
        Iterable<Angajat> angajati = getAllAngajati();
        for (Angajat angajat : angajati) {
            String username2 = angajat.getUsername();
            if (Objects.equals(username, username2))
                return angajat;
        }
        return null;
    }

    @Override
    public Iterable<Angajat> getAllAngajati() {
        return angajatRepo.getAll();
    }

    @Override
    public Spectacol getSpectacol(Long id){
        return spectacolRepo.findElement(id);
    }
    @Override
    public Spectacol getSpectacolByAll(String artistName, LocalTime time, String place){
        Iterable<Spectacol> spectacole = getAllSpectacole();
        for (Spectacol spectacol : spectacole) {
            String artistName2 = spectacol.getArtistName();
            LocalTime time2=spectacol.getDate().toLocalTime();
            String place2=spectacol.getPlace();
            if (Objects.equals(artistName, artistName2) && Objects.equals(time,time2)&&Objects.equals(place,place2))
                return spectacol;
        }
        return null;
    }
    @Override
    public Iterable<Spectacol> getAllSpectacole() {
        return spectacolRepo.getAll();
    }

    @Override
    public void addBilet(String cumparatorName, long idSpectacol, int nrSeats) throws MyException, ValidationException {
        if(getSpectacol(idSpectacol).getAvailableSeats()>=nrSeats)
        {
            Bilet toAdd = new Bilet(cumparatorName,idSpectacol,nrSeats);
            Spectacol spectacol=getSpectacol(idSpectacol);
            biletValidator.validate(toAdd);
            biletRepo.add(toAdd);
            spectacolRepo.update(spectacol,nrSeats);
            for (IObserver loggedUser: loggedAngajati.values()) {
                loggedUser.biletAdded(toAdd);
            }
        }
        else
        {
         throw new MyException("Nu exista atatea locuri disponibile.");
        }
    }

    public synchronized Angajat login(Angajat employee, IObserver client) throws Exception {
        Angajat employeeToLogin=angajatRepo.authenticateAngajat(employee.getUsername(),employee.getPassword());
        if (employeeToLogin!=null){
            if (loggedAngajati.get(employeeToLogin.getId())!=null)
                throw new Exception("Angajat already logged in.");
            loggedAngajati.put(employeeToLogin.getId(),client);
        }
        else
            throw new Exception("Authentication failed.");
        return employeeToLogin;
    }
    public synchronized void logout(Angajat employee, IObserver client) throws Exception {
        IObserver loggedEmployee= loggedAngajati.remove(employee.getId());
        if (loggedEmployee==null)
            throw new Exception("Employee "+String.valueOf(employee.getId())+" is not logged in.");
    }
}

File: Anul 2/Semestrul 2/MPP/labs/Tema 4-proiect Server-Client Csharp/proiectSCcsharp/server/Service.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using static System.Windows.Forms.VisualStyles.VisualStyleElement.StartPanel;
using System.Xml.Linq;
using System.Data.Entity.Core.Objects;
using proiectSCcsharp.domain.validators;
using proiectSCcsharp.domain;
using proiectSCcsharp.repository;
using proiectSCcsharp.services;

namespace proiectSCcsharp.server
{
    public class Service : IService
    {
        private readonly Validator<Angajat> angajatValidator;
        private readonly Validator<Spectacol> spectacolValidator;
        private readonly Validator<Bilet> biletValidator;
        private readonly RepoAngajat angajatRepo;
        private readonly RepoSpectacol spectacolRepo;
        private readonly RepoBilet biletRepo;
        private readonly IDictionary<long, IObserver> loggedAngajati;

        public Service(Validator<Angajat> angajatValidator, Validator<Spectacol> spectacolValidator, Validator<Bilet> biletValidator, RepoAngajat repoAngajat, RepoSpectacol repoSpectacol, RepoBilet repoBilet)
        {
            this.angajatValidator = angajatValidator;
            this.biletValidator = biletValidator;
            this.spectacolValidator = spectacolValidator;
            this.angajatRepo = repoAngajat;
            this.spectacolRepo = repoSpectacol;
            this.biletRepo = repoBilet;
            loggedAngajati = new Dictionary<long, IObserver>();
        }
        public Angajat getAngajatByUsername(String username)
        {
            IEnumerable<Angajat> angajati = getAllAngajati();
            foreach (Angajat angajat in angajati)
            {
                String username2 = angajat.username;
                if (username == username2)
                    return angajat;
            }
            return null;
        }

        public List<Angajat> getAllAngajati()
        {
            return angajatRepo.getAll();
        }

        public Spectacol getSpectacol(long id)
        {
            return spectacolRepo.findElement(id);
        }
        public Spectacol getSpectacolByAll(String artistName, String time, String place)
        {
            List <Spectacol> spectacole = getAllSpectacole();
            foreach (Spectacol spectacol in spectacole)
            {
                String artistName2 = spectacol.artistName;
                String time2 = spectacol.time;
                String place2 = spectacol.place;
                if (artistName==artistName2 && time ==time2 && place==place2)
                    return spectacol;
            }
            return null;
        }
        public List<Spectacol> getAllSpectacole()
        {
            return spectacolRepo.getAll();
        }
        
        public List<SpectacolDTO> getAllSpectacoleCautate(string date)
        {
            List<SpectacolDTO> temp=new List<SpectacolDTO> { };
            foreach(Spectacol spectacol in getAllSpectacole())
            {
                if(spectacol.date==date)
                {
                    SpectacolDTO spectacolDTO = new SpectacolDTO(spectacol.artistName,spectacol.time, spectacol.place, spectacol.availableSeats);
                    temp.Add(spectacolDTO);
                }
            }
            return temp;
        }
        public void addBilet(String cumparatorName, long idSpectacol, int nrSeats)
        {
            if (getSpectacol(idSpectacol).availableSeats>=nrSeats)
            {
                Bilet toAdd = new Bilet(cumparatorName, idSpectacol, nrSeats);
                Spectacol spectacol = getSpectacol(idSpectacol);
                biletRepo.add(toAdd);
                spectacolRepo.update(spectacol, nrSeats);
                foreach (var employeeO in loggedAngajati.Values)
                {
                    Task.Run(() => employeeO.BiletAdded(toAdd));
                }
            }
            else
            {
                throw new MyException("Nu exista atatea locuri disponibile.");
            }
        }

        public Angajat login(Angajat employee, IObserver employeeObserver)
        {
            var authenticateAngajat = angajatRepo.AuthenticateAngajat(employee.username, employee.password);
            if (authenticateAngajat != null)
            {
                if (loggedAngajati.ContainsKey(authenticateAngajat.id))
                    throw new MyException("User already logged in.");
                loggedAngajati[authenticateAngajat.id] = employeeObserver;
            }
            else
                throw new MyException("Authentication failed.");
            return authenticateAngajat;
        }

        public void logout(Angajat employee)
        {
            var localClient = loggedAngajati[employee.id];
            if (localClient == null)
                throw new MyException("User " + employee.id + " is not logged in.");
            loggedAngajati.Remove(employee.id);
        }
    }
}